component/xload/comp/splitter.go

```diff
@@ -0,0 +1,201 @@
+   copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE
+*/
+
+package comp
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/Azure/azure-storage-fuse/v2/common/log"
+	"github.com/Azure/azure-storage-fuse/v2/component/xload/common"
+	"github.com/Azure/azure-storage-fuse/v2/internal"
+)
+
+// verify that the below types implement the xcomponent interfaces
+var _ common.XComponent = &splitter{}
+var _ common.XComponent = &downloadSplitter{}
+
+const SPLITTER string = "splitter"
+
+type splitter struct {
+	common.XBase
+	blockSize uint64
+	blockPool *common.BlockPool
+	path      string
+}
+
+// --------------------------------------------------------------------------------------------------------
+
+type downloadSplitter struct {
+	splitter
+}
+
+func NewDownloadSplitter(blockSize uint64, blockPool *common.BlockPool, path string, remote internal.Component) (*downloadSplitter, error) {
+	log.Debug("splitter::NewDownloadSplitter : create new download splitter for %s, block size %v", path, blockSize)
+
+	d := &downloadSplitter{
+		splitter: splitter{
+			blockSize: blockSize,
+			blockPool: blockPool,
+			path:      path,
+		},
+	}
+
+	d.SetName(SPLITTER)
+	d.SetRemote(remote)
+	d.Init()
+	return d, nil
+}
+
+func (d *downloadSplitter) Init() {
+	d.SetThreadPool(common.NewThreadPool(common.MAX_DATA_SPLITTER, d.Process))
+	if d.GetThreadPool() == nil {
+		log.Err("downloadSplitter::Init : fail to init thread pool")
+	}
+}
+
+func (d *downloadSplitter) Start() {
+	log.Debug("downloadSplitter::Start : start download splitter for %s", d.path)
+	d.GetThreadPool().Start()
+}
+
+func (d *downloadSplitter) Stop() {
+	log.Debug("downloadSplitter::Stop : stop download splitter for %s", d.path)
+	if d.GetThreadPool() != nil {
+		d.GetThreadPool().Stop()
+	}
+	d.GetNext().Stop()
+}
+
+// download data in chunks and then write to the local file
+func (d *downloadSplitter) Process(item *common.WorkItem) (int, error) {
+	var err error
+
+	log.Debug("downloadSplitter::Process : Splitting data for %s", item.Path)
+	if len(item.Path) == 0 {
+		return 0, nil
+	}
+
+	numBlocks := ((item.DataLen - 1) / d.blockSize) + 1
+	offset := int64(0)
+
+	// TODO:: xload : should we delete the file if it already exists
+	// TODO:: xload : what should be the flags and mode and should we allocate the full size to the file
+	item.FileHandle, err = os.OpenFile(filepath.Join(d.path, item.Path), os.O_WRONLY|os.O_CREATE, 0644)
```

Comment:
Truncate the file to its actual size post creation so that if the disk is full, it fails here

Reply:
Done, doing truncate after file creation and before the download starts. If the disk is full, truncate operation will fail and return back, and the download of the file will not start.